addon/record-data.js

```diff
@@ -2,369 +2,479 @@
-        fragment.destroy();
-        delete this.fragments[key];
-      }
+
+  setDirtyFragment(key, value) {
+    const behavior = this._fragmentBehavior[key];
+    assert(`Attribute '${key}' for model '${this.modelName}' must be a fragment`, behavior != null);
+    const originalValue = key in this._inFlightFragments ? this._inFlightFragments[key] : this._fragmentData[key];
+    const isDirty = behavior.isDirty(value, originalValue);
+    const oldDirty = this.isFragmentDirty(key);
+    if (isDirty !== oldDirty) {
+      this.notifyStateChange(key);
+    }
+    if (isDirty) {
+      this._fragments[key] = value;
+      this.fragmentDidDirty();
+    } else {
+      delete this._fragments[key];
+      this.fragmentDidReset();
     }
+    // this._fragmentArrayCache[key]?.notify();
+  }
 
-    for (key in this.inFlightFragments) {
-      fragment = this.inFlightFragments[key];
-      if (fragment) {
-        fragment.destroy();
-        delete this.inFlightFragments[key];
+  setDirtyAttribute(key, value) {
+    assert(
+      `Attribute '${key}' for model '${this.modelName}' must not be a fragment`,
+      this._fragmentBehavior[key] == null
+    );
+    const oldDirty = this.isAttrDirty(key);
+    super.setDirtyAttribute(key, value);
+
+    const isDirty = this.isAttrDirty(key);
+    if (isDirty !== oldDirty) {
+      this.notifyStateChange(key);
+    }
+    if (isDirty) {
+      this.fragmentDidDirty();
+    } else {
+      this.fragmentDidReset();
+    }
+  }
+
+  getFragmentOwner() {
+    return this._fragmentOwner?.recordData;
+  }
+
+  setFragmentOwner(recordData, key) {
+    assert('Fragment owner must be a RecordData', recordData instanceof RecordData);
+    assert('Fragment owner key must be a fragment', recordData._fragmentBehavior[key] != null);
+    assert(
+      'Fragments can only belong to one owner, try copying instead',
+      !this._fragmentOwner || this._fragmentOwner.recordData === recordData
+    );
+    this._fragmentOwner = { recordData, key };
+  }
+
+  _newFragmentRecordData(definition, attributes) {
+    const type = getActualFragmentType(definition.modelName, definition.options, attributes);
+    const recordData = this.storeWrapper.recordDataFor(type);
+    recordData.setFragmentOwner(this, definition.name);
+    recordData._fragmentPushData({ attributes });
+    return recordData;
+  }
+
+  hasChangedAttributes() {
+    return super.hasChangedAttributes() || this.hasChangedFragments();
+  }
+
+  hasChangedFragments() {
+    return this.__fragments !== null && Object.keys(this.__fragments).length > 0;
+  }
+
+  isFragmentDirty(key) {
+    return this.__fragments?.[key] !== undefined;
+  }
+
+  getCanonicalState() {
+    const result = Object.assign({}, this._data);
+    for (const [key, behavior] of Object.entries(this._fragmentBehavior)) {
+      if (!this.hasFragment(key)) {
+        // force the fragment attribute to initialize its default value
+        internalModelFor(this).getRecord().get(key);
```

Comment:
This sucks. We can't serialize fragment state without involving the fragment attributes which are responsible for default values.

Reply:
Yeah this is something I want to deprecate in ember-data too.

Reply:
intenralModelFor helper will need adjusted for 4.5, that said I don't think you need it. You have access to the store and the identifier, just peek the record.